test/unit/asrouter/CFRPageActions.test.js

```diff
@@ -0,0 +1,634 @@
+        global.gBrowser.selectedBrowser = {}; // Some other browser
+        assert.isFalse(await CFRPageActions.addRecommendation(fakeBrowser, fakeHost, fakeRecommendation, dispatchStub));
+      });
+      it("should fail and not add a recommendation if the host doesn't match", async () => {
+        const someOtherFakeHost = "subdomain.mozilla.com";
+        assert.isFalse(await CFRPageActions.addRecommendation(fakeBrowser, someOtherFakeHost, fakeRecommendation, dispatchStub));
+      });
+      it("should otherwise succeed and add an element to the RecommendationMap", async () => {
+        assert.isTrue(await CFRPageActions.addRecommendation(fakeBrowser, fakeHost, fakeRecommendation, dispatchStub));
+        assert.deepInclude(CFRPageActions.RecommendationMap.get(fakeBrowser), {
+          id: fakeRecommendation.id,
+          host: fakeHost,
+          content: fakeRecommendation.content
+        });
+      });
+      it("should create a PageAction if one doesn't exist for the window, save it in the PageActionMap, and call `show`", async () => {
+        const win = fakeBrowser.ownerGlobal;
+        assert.isFalse(CFRPageActions.PageActionMap.has(win));
+        await CFRPageActions.addRecommendation(fakeBrowser, fakeHost, fakeRecommendation, dispatchStub);
+        const pageAction = CFRPageActions.PageActionMap.get(win);
+        assert.equal(win, pageAction.window);
+        assert.equal(dispatchStub, pageAction._dispatchToASRouter);
+        assert.calledOnce(PageAction.prototype.show);
+      });
+    });
+
+    describe("clearRecommendations", () => {
+      const createFakePageAction = () => ({hide: sandbox.stub()});
+      const windows = [{}, {}, {closed: true}];
+      const browsers = [{}, {}, {}, {}];
+
+      beforeEach(() => {
+        CFRPageActions.PageActionMap.set(windows[0], createFakePageAction());
+        CFRPageActions.PageActionMap.set(windows[2], createFakePageAction());
+        for (const browser of browsers) {
+          CFRPageActions.RecommendationMap.set(browser, {});
+        }
+        globals.set({Services: {wm: {getEnumerator: () => windows}}});
+      });
+
+      it("should hide the PageActions of any existing, non-closed windows", () => {
+        const pageActions = windows.map(win => CFRPageActions.PageActionMap.get(win));
+        CFRPageActions.clearRecommendations();
+
+        // Only the first window had a PageAction and wasn't closed
+        assert.calledOnce(pageActions[0].hide);
+        assert.isUndefined(pageActions[1]);
+        assert.notCalled(pageActions[2].hide);
+      });
+      it("should clear the PageActionMap and the RecommendationMap", () => {
+        CFRPageActions.clearRecommendations();
+
+        // Both are WeakMaps and so are not iterable, cannot be cleared, and
+        // cannot have their length queried directly, so we have to check
+        // whether previous elements still exist
+        for (const win of windows) {
```

Comment:
This also needs some assertion related to the number of `windows` or `browsers` we expect to find otherwise (if 0) the test doesn't assert anything.